File: weather-demo-app-apple/Weather/Data/WeatherRepository.swift
//
//  WeatherRepository.swift
//  Weather
//
//  Created by Onur on 4/11/26.
//

import Foundation

final class WeatherRepository: WeatherDataSource {
    private let api: WeatherApi
    private let preferences: PreferencesDataSource

    init(
        api: WeatherApi,
        preferences: PreferencesDataSource
    ) {
        self.api = api
        self.preferences = preferences
    }

    func getCurrentWeather(city: String) async throws -> CurrentWeather {
        do {
            let normalizedCity = try city.normalizedCity()

            let result = try await api.getCurrentWeather(
                city: normalizedCity
            ).toDomain()

            preferences.saveLastSearchedCity(result.city)
            return result

        } catch {
            throw error.toRepositoryError()
        }
    }

    func getForecast(city: String) async throws -> [ForecastDay] {
        let normalizedCity = try city.normalizedCity()

        return try await api.getForecast(
            city: normalizedCity
        ).toDomainDaily()
    }

    func saveLastSearchedCity(_ city: String) {
        preferences.saveLastSearchedCity(
            city.trimmingCharacters(in: .whitespacesAndNewlines)
        )
    }

    func getLastSearchedCity() -> String? {
        return preferences.getLastSearchedCity()
    }

    func setWelcomeSeen(_ seen: Bool) {
        preferences.setWelcomeSeen(seen)
    }

    func hasSeenWelcome() -> Bool {
        return preferences.hasSeenWelcome()
    }
}

File: weather-demo-app-android/app/src/main/java/com/demo/weather/data/WeatherRepository.kt
package com.demo.weather.data

import com.demo.weather.domain.model.CurrentWeather
import com.demo.weather.domain.model.ForecastDay
import com.demo.weather.extension.toDomain
import com.demo.weather.extension.toDomainDaily
import com.demo.weather.extension.normalizedCity
import com.demo.weather.extension.toRepositoryError
import com.demo.weather.network.WeatherApi

class WeatherRepository(
    private val api: WeatherApi,
    private val preferences: PreferencesDataSource
) : WeatherDataSource {

    override suspend fun getCurrentWeather(city: String): CurrentWeather {
        return try {
            val normalizedCity = city.normalizedCity()

            val result = api.getCurrentWeather(
                city = normalizedCity
            ).toDomain()

            preferences.saveLastSearchedCity(result.city)
            result

        } catch (t: Throwable) {
            throw t.toRepositoryError()
        }
    }

    override suspend fun getForecast(city: String): List<ForecastDay> {
        return try {
            val normalizedCity = city.normalizedCity()

            api.getForecast(
                city = normalizedCity
            ).toDomainDaily()

        } catch (t: Throwable) {
            throw t.toRepositoryError()
        }
    }

    override fun saveLastSearchedCity(city: String) {
        preferences.saveLastSearchedCity(city.trim())
    }

    override fun getLastSearchedCity(): String? {
        return preferences.getLastSearchedCity()
    }

    override fun setWelcomeSeen(seen: Boolean) {
        preferences.setWelcomeSeen(seen)
    }

    override fun hasSeenWelcome(): Boolean {
        return preferences.hasSeenWelcome()
    }
}